nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -330,6 +351,123 @@ impl DataStore {
+    /// [`InstanceSnapshot`], which contains the following:
+    ///
+    /// - The [`Instance`] record itself,
+    /// - The instance's active [`Vmm`] record, if the `active_propolis_id`
+    ///   column is not null,
+    /// - The instance's target [`Vmm`] record, if the `target_propolis_id`
+    ///   column is not null,
+    /// - The instance's current active [`Migration`], if the `migration_id`
+    ///   column is not null.
+    pub async fn instance_fetch_all(
+        &self,
+        opctx: &OpContext,
+        authz_instance: &authz::Instance,
+    ) -> LookupResult<InstanceSnapshot> {
+        opctx.authorize(authz::Action::Read, authz_instance).await?;
+
+        use db::schema::instance::dsl as instance_dsl;
+        use db::schema::migration::dsl as migration_dsl;
+        use db::schema::vmm::dsl as vmm_dsl;
+
+        let mut results = instance_dsl::instance
+            .filter(instance_dsl::id.eq(authz_instance.id()))
+            .filter(instance_dsl::time_deleted.is_null())
+            .left_join(
+                vmm_dsl::vmm.on((vmm_dsl::id
+                    .nullable()
+                    .eq(instance_dsl::active_propolis_id)
+                    .or(vmm_dsl::id
+                        .nullable()
+                        .eq(instance_dsl::target_propolis_id)))
+                .and(vmm_dsl::time_deleted.is_null())),
+            )
+            .left_join(
+                migration_dsl::migration.on(migration_dsl::id
+                    .nullable()
+                    .eq(instance_dsl::migration_id)
+                    .and(migration_dsl::time_deleted.is_null())),
+            )
+            .select((
+                Instance::as_select(),
+                Option::<Vmm>::as_select(),
+                Option::<Migration>::as_select(),
+            ))
+            .load_async::<(Instance, Option<Vmm>, Option<Migration>)>(
+                &*self.pool_connection_authorized(opctx).await?,
+            )
+            .await
+            .map_err(|e| {
+                public_error_from_diesel(
+                    e,
+                    ErrorHandler::NotFoundByLookup(
+                        ResourceType::Instance,
+                        LookupType::ById(authz_instance.id()),
+                    ),
+                )
+            })?
+            .into_iter();
+
+        // Because we join with the `vmm` table on VMMs whose ID is the active
+        // *or* target VMM ID in the instance record, this query may return a
+        // result set with multiple rows. If it does, we have to iterate over
+        // them and determine which VMM result is the active VMM and which is the
+        // target. The instance and migration records should be identical, so we
+        // only need the first ones.
+        //
+        // Yes, this code is a bit unfortunate, but Diesel doesn't like the idea
+        // of joining with the same table twice on different columns...so, this,
+        // at least, works.
```

Comment:
Yeah, this is probably the best way of doing this while coloring inside the lines Diesel has set out. (I could imagine having some kind of CTE that conjures up a temporary table that contains the VMM IDs and a "sequence" column that can be used to order the returned rows unambiguously, but I've found it very hard to do this kind of thing without dropping out to raw SQL.)

Reply:
That's sort of what I thought. Writing the CTE felt like it would be enough of a pain that I didn't really think it would be worth the effort, but we can revisit this later.